File: src/main/java/com/jcloisterzone/board/pointer/ScoreMeeplePositionsPointer.java
package com.jcloisterzone.board.pointer;

import com.jcloisterzone.Immutable;
import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.MeeplePointer;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.figure.Meeple;

import io.vavr.collection.Set;
import io.vavr.Tuple2;

/**
 * Points on feature on board or placed meeple.
 */
@Immutable
public class ScoreMeeplePositionsPointer implements BoardPointer {

    private static final long serialVersionUID = 1L;

    private final FeaturePointer featurePointer;
    private final String meepleId;
    private final Set<Position> positions;

    public ScoreMeeplePositionsPointer(FeaturePointer featurePointer, String meepleId, Set<Position> positions) {
        this.featurePointer = featurePointer;
        this.meepleId = meepleId;
        this.positions = positions;
    }

    public FeaturePointer asFeaturePointer() {
        return featurePointer;
    }

    public Position getPosition() {
        return featurePointer.getPosition();
    }

    public Location getLocation() {
        return featurePointer.getLocation();
    }

    public String getMeepleId() {
        return meepleId;
    }

    public Set<Position> getPositions() {
        return positions;
    }

    @Override
    public String toString() {
        return "[x=" + getPosition().x + ",y=" + getPosition().y + "," + getLocation() + "," + this.getMeepleId() + ",positions=" + positions.mkString(",") + "]";
    }
}

File: src/main/ts/com/jcloisterzone/board/pointer/ScoreMeeplePositionsPointer.ts
import type { Set } from "../../../../io/vavr/Set.js";
import type { Location } from "../Location.js";
import type { Position } from "../Position.js";
import type { BoardPointer } from "./BoardPointer.js";
import type { FeaturePointer } from "./FeaturePointer.js";

/** A scoring source pointing at a placed meeple plus the set of positions it scored
 *  over (used by the obelisk's range scoring). */
export class ScoreMeeplePositionsPointer implements BoardPointer {
  constructor(
    private readonly featurePointer: FeaturePointer,
    private readonly meepleId: string,
    private readonly positions: Set<Position>,
  ) {}

  asFeaturePointer(): FeaturePointer {
    return this.featurePointer;
  }

  getPosition(): Position {
    return this.featurePointer.getPosition();
  }

  getLocation(): Location | null {
    return this.featurePointer.getLocation();
  }

  getMeepleId(): string {
    return this.meepleId;
  }

  getPositions(): Set<Position> {
    return this.positions;
  }

  toString(): string {
    const p = this.getPosition();
    return `[x=${p.x},y=${p.y},${this.getLocation()},${this.meepleId},positions=${this.positions.mkString(",")}]`;
  }
}
